services/blockchain-connector/shared/sdk/client.js

```diff
@@ -186,44 +130,83 @@ const invokeEndpoint = async (endpoint, params = {}, numRetries = NUM_REQUEST_RE
-		if (
-			!isGenesisBlockIndexed &&
-			indexStatus.data &&
-			indexStatus.data.genesisHeight <= indexStatus.data.lastIndexedBlockHeight
-		) {
-			clearInterval(intervalTimeout);
-			triggerRegularClientLivelinessChecks(CLIENT_ALIVE_ASSUMPTION_TIME);
-			isGenesisBlockIndexed = true;
-			logger.info(
-				`API client heartbeat checks re-scheduled to run every ${CLIENT_ALIVE_ASSUMPTION_TIME}ms.`,
-			);
+const disconnectClient = async cachedClient => {
+	try {
+		await cachedClient.disconnect();
+	} catch (err) {
+		logger.warn(`Client disconnection failed due to: ${err.message}`);
+	}
+};
+
+const refreshClientsCache = async () => {
+	// Indicates if an active client was destroyed or no active clients available
+	let activeClientNotAvailable = cachedApiClients.length === 0;
+
+	// Check liveliness and remove non-responsive clients
+	let index = 0;
+	while (index < cachedApiClients.length) {
+		const cachedClient = cachedApiClients[index];
+		try {
+			if (!(await checkIsClientAlive(cachedClient))) {
+				cachedApiClients.splice(index, 1);
+				if (index === 0) {
+					activeClientNotAvailable = true;
+				}
+				await disconnectClient(cachedClient);
+			} else {
+				index++;
+			}
+		} catch (err) {
+			logger.info(`Failed to check client aliveness.\nError:${err.message}`);
+		}
+	}
+
+	// Initiate new clients if necessary
+	let missingClientCount = CONNECTION_LIMIT - cachedApiClients.length;
+
+	while (missingClientCount-- > 0) {
+		try {
+			await instantiateAndCacheClient();
+		} catch (err) {
+			logger.info(`Failed to instantiate new api client.\nError:${err.message}`);
 		}
-	};
+	}
 
-	Signals.get('genesisBlockDownloaded').add(genesisBlockDownloadedListener);
-	Signals.get('updateIndexStatus').add(genesisBlockIndexedListener);
-}
+	// Reset event listeners if active api client was destroyed
+	if (activeClientNotAvailable && cachedApiClients.length > 0) {
+		Signals.get('newApiClient').dispatch();
+	}
+
+	return cachedApiClients.length;
+};
+
+// Listen to client unresponsive events and try to reinitialize client
+const resetApiClientListener = async () => {
+	logger.info(
+		`Received API client reset signal. Will drop ${cachedApiClients.length} cached API client(s).`,
+	);
+	// Drop pool of cached clients. Will be instantiated by refresh clients job
+	while (cachedApiClients.length > 0) {
+		const cachedClient = cachedApiClients.pop();
+		await disconnectClient(cachedClient);
+	}
+};
+Signals.get('resetApiClient').add(resetApiClientListener);
+
+// Check periodically for client aliveness and refill cached clients pool
+(async () => {
+	// eslint-disable-next-line no-constant-condition
+	while (true) {
+		const cacheRefreshStartTime = Date.now();
+		await refreshClientsCache();
+		logger.debug(
+			`Refreshed api client cached in ${Date.now() - cacheRefreshStartTime}ms. There are ${
+				cachedApiClients.length
+			} API client(s) in the pool.`,
+		);
+		await delay(CLIENT_ALIVE_ASSUMPTION_TIME);
+	}
+})();
```

Comment:
Better to create a molecular job than an infinite while loop?

Reply:
True, it would make code more neat. But one issue with jobs is it would use interval which does not wait for current round of checks to finish and we may endup having multiple executions of `refreshClientsCache` at same time. This would complicate the logic with locks and other checks.